jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -0,0 +1,510 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import org.eclipse.jetty.io.content.ByteBufferContentSource;
+import org.eclipse.jetty.io.content.InputStreamContentSource;
+import org.eclipse.jetty.io.content.PathContentSource;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.IteratingNestedCallback;
+import org.eclipse.jetty.util.resource.MemoryResource;
+import org.eclipse.jetty.util.resource.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Common IO operations for {@link Resource} content.
+ */
+public class IOResources
+{
+    /**
+     * <p>Reads the contents of a Resource into a RetainableByteBuffer.</p>
+     * <p>The resource must not be a directory, must exists and there must be
+     * a way to access its contents.</p>
+     * <p>Multiple optimized methods are used to access the resource's contents but if they all fail,
+     * {@link Resource#newInputStream()} is used as a fallback.</p>
+     *
+     * @param resource the resource to be read.
+     * @param bufferPool the {@link ByteBufferPool} to get buffers from. null means allocate new buffers as needed.
+     * @param direct the directness of the buffers, this parameter is ignored if {@code bufferSize} is &lt; 1.
+     * @return a {@link RetainableByteBuffer} containing the resource's contents.
+     * @throws IllegalArgumentException if the resource is a directory or does not exist or there is no way to access its contents.
+     */
+    public static RetainableByteBuffer toRetainableByteBuffer(Resource resource, ByteBufferPool bufferPool, boolean direct) throws IllegalArgumentException
```

Comment:
The only non-test usage of this method is in `CachingHttpContentFactory` where we want to keep an in-memory copy of the resource for caching purposes.

I'm not sure how the RBB changes would help here?

Reply:
I'm just saying that if there was a need, then we could hide a whole resource reader within the RBB abstraction, and so long as nobody called getByteBuffer, then we'd never have the whole resource in memory.     I.e. the power of using RBB as the primary interface rather than RBB.getByteBuffer